File: src/main/java/com/weeth/domain/attendance/domain/service/AttendanceScheduler.java
package com.weeth.domain.attendance.domain.service;

import jakarta.transaction.Transactional;
import java.util.List;
import com.weeth.domain.attendance.domain.entity.Attendance;
import com.weeth.domain.attendance.domain.entity.Session;
import com.weeth.domain.schedule.domain.service.MeetingGetService;
import lombok.RequiredArgsConstructor;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class AttendanceScheduler {

    private final MeetingGetService meetingGetService;
    private final AttendanceGetService attendanceGetService;
    private final AttendanceUpdateService attendanceUpdateService;

    @Transactional
    @Scheduled(cron = "0 0 22 * * THU", zone = "Asia/Seoul")
    public void autoCloseAttendance() {
        List<Session> sessions = meetingGetService.findAllOpenMeetingsBeforeNow();

        sessions.forEach(session -> {
            session.close();
            List<Attendance> attendanceList = attendanceGetService.findAllByMeeting(session);
            attendanceUpdateService.close(attendanceList);
        });
    }
}

File: src/main/kotlin/com/weeth/domain/attendance/domain/service/AttendanceScheduler.kt
package com.weeth.domain.attendance.domain.service

import com.weeth.domain.schedule.domain.service.MeetingGetService
import jakarta.transaction.Transactional
import org.springframework.scheduling.annotation.Scheduled
import org.springframework.stereotype.Service

@Service
class AttendanceScheduler(
    private val meetingGetService: MeetingGetService,
    private val attendanceGetService: AttendanceGetService,
    private val attendanceUpdateService: AttendanceUpdateService,
) {
    @Transactional
    @Scheduled(cron = "0 0 22 * * THU", zone = "Asia/Seoul")
    fun autoCloseAttendance() {
        val meetings = meetingGetService.findAllOpenMeetingsBeforeNow()

        meetings.forEach { meeting ->
            meeting.close()
            val attendanceList = attendanceGetService.findAllByMeeting(meeting)
            attendanceUpdateService.close(attendanceList)
        }
    }
}
